File: js/src/utils/nvidia.ts
import { Buffer } from 'buffer';
import { createLocalJWKSet, errors, jwtVerify } from 'jose';
import {
  decodeNrasOverallAttestationClaims,
  decodeNrasOverallAttestationJwt,
  decodeNvidiaJwks,
  decodeNvidiaPayloadNonce,
} from '../boundaries/nvidia';
import type {
  CreateGpuEvidenceVerifierParams,
  GpuEvidenceVerifier,
} from '../types/verification';
import { NVIDIA_GPU_VERIFIER_API_URL } from './consts';
import { isVerificationError, VerificationError } from './errors';
import type { VerificationFailure } from './errors';

type NvidiaJwksResolver = ReturnType<typeof createLocalJWKSet>;
type NvidiaJwtFailureReason = Extract<
  VerificationFailure,
  {
    code: 'gpu.jwt_verification_failed';
  }
>['details']['reason'];

const NVIDIA_ISSUER = 'https://nras.attestation.nvidia.com';
const NVIDIA_JWKS_URL = `${NVIDIA_ISSUER}/.well-known/jwks.json`;

/**
 * Use NVIDIA's signed verdict through NRAS or a forwarding proxy. The JWT
 * remains bound to the submitted payload nonce. A custom JWKS URL must be a
 * trusted source of NVIDIA's public keys; it does not change the expected issuer.
 * Model verification separately binds that payload nonce to the client nonce.
 */
export function createGpuEvidenceVerifier({
  nrasUrl = NVIDIA_GPU_VERIFIER_API_URL,
  jwksUrl = NVIDIA_JWKS_URL,
}: CreateGpuEvidenceVerifierParams = {}): GpuEvidenceVerifier {
  return async (nvidiaPayload) =>
    nvidiaNrasVerifier({
      nvidiaPayload,
      nonce: decodeNvidiaPayloadNonce(nvidiaPayload),
      nrasUrl,
      jwksUrl,
    });
}

type NvidiaNrasVerifierParams = {
  nvidiaPayload: string;
  nonce: string;
  nrasUrl: string;
  jwksUrl: string;
};

/**
 * Verify NRAS's overall EAT JWT using NVIDIA's public JWKS. Only ES384 is
 * accepted, matching NVIDIA's remote verifier. Detached device claims are not
 * consumed here; the signed overall result is the SDK's GPU verdict.
 */
async function nvidiaNrasVerifier({
  nvidiaPayload,
  nonce,
  nrasUrl,
  jwksUrl,
}: NvidiaNrasVerifierParams): Promise<void> {
  const response = await fetchNras(nvidiaPayload, nrasUrl);

  if (!response.ok) {
    throw new VerificationError({
      code: 'gpu.nras_request_failed',
      details: { reason: 'http_status', status: response.status },
      retryable: isRetryableNrasStatus(response.status),
    });
  }

  const raw = await getNrasJson(response);
  const token = decodeNrasOverallAttestationJwt(raw);
  const jwks = await fetchNvidiaJwks(jwksUrl);
  let payload: unknown;
  try {
    const verified = await jwtVerify(
      token,
      (header, signature) => {
        if (header.kid === undefined) throw jwtFailure('key_not_found');
        return jwks(header, signature);
      },
      {
        algorithms: ['ES384'],
        issuer: NVIDIA_ISSUER,
        requiredClaims: ['exp', 'nbf', 'iat'],
      },
    );
    payload = verified.payload;
  } catch (cause) {
    if (isVerificationError(cause)) throw cause;
    if (cause instanceof errors.JWTExpired) throw jwtFailure('expired', cause);
    if (cause instanceof errors.JWTClaimValidationFailed) {
      throw jwtFailure(
        cause.claim === 'nbf' && cause.reason === 'check_failed'
          ? 'not_yet_valid'
          : 'invalid_claims',
        cause,
      );
    }
    if (cause instanceof errors.JWTInvalid)
      throw jwtFailure('invalid_claims', cause);
    if (cause instanceof errors.JWKSNoMatchingKey)
      throw jwtFailure('key_not_found', cause);
    if (cause instanceof errors.JOSEAlgNotAllowed)
      throw jwtFailure('unsupported_algorithm', cause);
    throw jwtFailure('invalid_signature', cause);
  }
  const claims = decodeNrasOverallAttestationClaims(payload);
  if (claims.iat > Date.now() / 1000) throw jwtFailure('not_yet_valid');
  const expectedNonce = nonce.replace(/^0x/i, '');
  if (
    !Buffer.from(claims.eat_nonce, 'hex').equals(
      Buffer.from(expectedNonce, 'hex'),
    )
  ) {
    throw jwtFailure('nonce_mismatch');
  }
  if (claims['x-nvidia-overall-att-result']) {
    return;
  }
  throw new VerificationError({
    code: 'gpu.attestation_rejected',
    details: { source: 'nras' },
  });
}

async function fetchNvidiaJwks(jwksUrl: string): Promise<NvidiaJwksResolver> {
  let response: Response;
  try {
    response = await fetch(jwksUrl);
  } catch (cause) {
    throw new VerificationError(
      {
        code: 'gpu.jwks_request_failed',
        details: { reason: 'transport' },
        retryable: true,
      },
      { cause },
    );
  }
  if (!response.ok) {
    throw new VerificationError({
      code: 'gpu.jwks_request_failed',
      details: { reason: 'http_status', status: response.status },
      retryable: isRetryableNrasStatus(response.status),
    });
  }
  try {
    const raw: unknown = await response.json();
    return createLocalJWKSet(decodeNvidiaJwks(raw));
  } catch (cause) {
    throw new VerificationError(
      {
        code: 'gpu.nras_response_invalid',
        details: { reason: 'invalid_jwks' },
      },
      { cause },
    );
  }
}

function jwtFailure(
  reason: NvidiaJwtFailureReason,
  cause?: unknown,
): VerificationError {
  return new VerificationError(
    {
      code: 'gpu.jwt_verification_failed',
      details: { reason },
    },
    cause === undefined ? undefined : { cause },
  );
}

async function fetchNras(
  nvidiaPayload: string,
  nrasUrl: string,
): Promise<Response> {
  try {
    return await fetch(nrasUrl, {
      method: 'POST',
      headers: {
        'content-type': 'application/json',
      },
      body: nvidiaPayload,
    });
  } catch (cause) {
    throw new VerificationError(
      {
        code: 'gpu.nras_request_failed',
        details: { reason: 'transport' },
        retryable: true,
      },
      { cause },
    );
  }
}

async function getNrasJson(response: Response): Promise<unknown> {
  try {
    return await response.json();
  } catch (cause) {
    throw invalidNrasResponse('invalid_json', cause);
  }
}

function invalidNrasResponse(
  reason: 'invalid_json',
  cause?: unknown,
): VerificationError {
  return new VerificationError(
    {
      code: 'gpu.nras_response_invalid',
      details: { reason },
    },
    cause === undefined ? undefined : { cause },
  );
}

function isRetryableNrasStatus(status: number): boolean {
  return status === 408 || status === 429 || status >= 500;
}

File: py/src/nearai_inference_sdk/utils/nvidia.py
"""Default NVIDIA NRAS evidence adapter."""

from __future__ import annotations

import json

import aiohttp
import jwt
from pydantic import ValidationError

from ..schemas import (
    NvidiaPayloadNonceSchema,
    _NrasOverallAttestationJwtClaimsSchema,
    _NrasResponseSchema,
    _NvidiaJwksSchema,
)
from ..types.verification import NvidiaEvidenceVerifier
from .common import require_byte_length
from .consts import NVIDIA_GPU_VERIFIER_API_URL
from .errors import verification_failure
from .fetch import FetchResponse, fetch

NVIDIA_ISSUER = 'https://nras.attestation.nvidia.com'
NVIDIA_JWKS_URL = f'{NVIDIA_ISSUER}/.well-known/jwks.json'


def create_nvidia_evidence_verifier(
    nras_url: str = NVIDIA_GPU_VERIFIER_API_URL,
    jwks_url: str = NVIDIA_JWKS_URL,
) -> NvidiaEvidenceVerifier:
    """Create the NVIDIA verifier with NRAS and signing-key URLs.

    The issuer remains NVIDIA. The JWKS endpoint supplies trusted signing keys;
    configure only a trusted source. The returned verifier binds the signed
    verdict to the payload nonce; callers must bind that nonce to their request,
    as verify_model_attestation does.
    """

    async def verify_evidence(payload: str) -> None:
        nonce = decode_nvidia_payload_nonce(payload)
        await verify_nvidia_nras(payload, nonce, nras_url, jwks_url)

    return verify_evidence


def decode_nvidia_payload_nonce(payload: str) -> str:
    """Decode GPU evidence and require a 32-byte hexadecimal nonce."""

    try:
        parsed = NvidiaPayloadNonceSchema.model_validate_json(payload)
    except ValidationError as error:
        reason = (
            'invalid_json'
            if error.errors(include_url=False)[0]['type'] == 'json_invalid'
            else 'nonce_missing'
        )
        raise verification_failure(
            'gpu.payload_invalid', {'reason': reason}, cause=error
        ) from error
    require_byte_length(parsed.nonce, 32, 'nvidia_payload.nonce')
    return parsed.nonce


async def verify_nvidia_nras(
    nvidia_payload: str,
    nonce: str,
    nras_url: str = NVIDIA_GPU_VERIFIER_API_URL,
    jwks_url: str = NVIDIA_JWKS_URL,
) -> None:
    """Verify NRAS's ES384-signed overall verdict, issuer, time and nonce.

    Detached device claims are not consumed by this adapter.
    """

    try:
        response = await fetch(
            nras_url,
            method='POST',
            data=nvidia_payload,
            headers={'content-type': 'application/json'},
        )
    except (aiohttp.ClientError, TimeoutError, OSError) as error:
        raise verification_failure(
            'gpu.nras_request_failed',
            {'reason': 'transport'},
            retryable=True,
            cause=error,
        ) from error

    if not response.ok:
        raise verification_failure(
            'gpu.nras_request_failed',
            {'reason': 'http_status', 'status': response.status},
            retryable=response.status in {408, 429} or response.status >= 500,
        )

    raw = _decode_nras_json(response)
    token = _decode_nras_overall_attestation_jwt(raw)
    key_set = await _fetch_nvidia_jwks(jwks_url)
    claims = _verify_nras_jwt(token, key_set)
    if bytes.fromhex(claims.eat_nonce) != bytes.fromhex(
        nonce.removeprefix('0x').removeprefix('0X')
    ):
        raise _jwt_failure('nonce_mismatch')
    if claims.overall_attestation_result:
        return
    raise verification_failure('gpu.attestation_rejected', {'source': 'nras'})


def _decode_nras_json(response: FetchResponse) -> object:
    try:
        return json.loads(response.text())
    except (UnicodeDecodeError, json.JSONDecodeError) as error:
        raise _invalid_nras_response('invalid_json', error) from error


def _decode_nras_overall_attestation_jwt(value: object) -> str:
    try:
        return _NrasResponseSchema.model_validate(value).overall_attestation_jwt
    except ValidationError as error:
        raise _invalid_nras_response('invalid_schema', error) from error


async def _fetch_nvidia_jwks(jwks_url: str) -> jwt.PyJWKSet:
    try:
        response = await fetch(jwks_url)
    except (aiohttp.ClientError, TimeoutError, OSError) as error:
        raise verification_failure(
            'gpu.jwks_request_failed',
            {'reason': 'transport'},
            retryable=True,
            cause=error,
        ) from error
    if not response.ok:
        raise verification_failure(
            'gpu.jwks_request_failed',
            {'reason': 'http_status', 'status': response.status},
            retryable=response.status in {408, 429} or response.status >= 500,
        )
    try:
        key_set = _NvidiaJwksSchema.model_validate_json(response.text())
        return jwt.PyJWKSet.from_dict(key_set.model_dump())
    except (jwt.PyJWTError, ValueError, TypeError, KeyError) as error:
        raise _invalid_nras_response('invalid_jwks', error) from error


def _verify_nras_jwt(
    token: str, key_set: jwt.PyJWKSet
) -> _NrasOverallAttestationJwtClaimsSchema:
    try:
        header = jwt.get_unverified_header(token)
        if header.get('alg') != 'ES384':
            raise _jwt_failure('unsupported_algorithm')
        kid = header.get('kid')
        if not isinstance(kid, str):
            raise _jwt_failure('key_not_found')
        try:
            key = key_set[kid]
        except KeyError as error:
            raise _jwt_failure('key_not_found', error) from error
        value = jwt.decode(
            token,
            key.key,
            algorithms=['ES384'],
            issuer=NVIDIA_ISSUER,
            options={'require': ['exp', 'nbf', 'iat'], 'verify_aud': False},
        )
        return _NrasOverallAttestationJwtClaimsSchema.model_validate(value)
    except jwt.ExpiredSignatureError as error:
        raise _jwt_failure('expired', error) from error
    except jwt.ImmatureSignatureError as error:
        raise _jwt_failure('not_yet_valid', error) from error
    except jwt.InvalidSignatureError as error:
        raise _jwt_failure('invalid_signature', error) from error
    except jwt.InvalidAlgorithmError as error:
        raise _jwt_failure('unsupported_algorithm', error) from error
    except jwt.InvalidTokenError as error:
        raise _jwt_failure('invalid_claims', error) from error
    except jwt.PyJWTError as error:
        raise _jwt_failure('invalid_signature', error) from error
    except (ValidationError, TypeError) as error:
        raise _jwt_failure('invalid_claims', error) from error


def _jwt_failure(reason: str, cause: BaseException | None = None):
    return verification_failure(
        'gpu.jwt_verification_failed', {'reason': reason}, cause=cause
    )


def _invalid_nras_response(reason: str, cause: BaseException | None = None):
    return verification_failure(
        'gpu.nras_response_invalid', {'reason': reason}, cause=cause
    )
